File: Permutations/permutations.js
const checkPermutations = (result, permutation, nums) => {
  if (permutation.length === nums.length) return result.push([...permutation]);
  for (let i = 0; i < nums.length; i++) {
    if (permutation.includes(nums[i])) continue;
    permutation.push(nums[i]);
    checkPermutations(result, permutation, nums);
    permutation.pop();
  }
};

const permute = (nums) => {
  const result = [];
  checkPermutations(result, [], nums);
  return result;
};

// console.log(permute([1, 2, 3]));
// console.log(permute([0, 1]));
// console.log(permute([1]));

File: Permutations/Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        results = []
        subset = []

        def dfs():
            if len(subset) == n:
                results.append(subset.copy())
                return

            for num in nums:
                if num not in subset:
                    subset.append(num)
                    dfs()
                    subset.pop()

        dfs()
        return results